internal/pkg/config/matcher_test.go

```diff
@@ -0,0 +1,244 @@
+package config
+
+import (
+	"testing"
+
+	"sigs.k8s.io/yaml"
+)
+
+func TestFuncMatcher(t *testing.T) {
+	testCases := []struct {
+		desc, yaml        string
+		path, recv, name  string
+		shouldErrorOnLoad bool
+		shouldMatch       bool
+	}{
+		{
+			desc: "Garbage in garbage out",
+			yaml: `
+Blahblah: foo
+PackageRE: bar`,
+			// TODO Discuss loss of strict loading
+			shouldErrorOnLoad: false, // TODO true
+		},
```

Comment:
I have discovered an interesting side-effect of using the K8s `yaml` package.

To massage some other pain points, that `yaml` package converts the YAML is it unmarshaling into JSON, and uses `json.Unmarshal` to unpack it.  However, the `json` package doesn't have a "strict" mode.  So when providing our own `UnmarshalJSON` methods, there is no strict mode inherited from our original `yaml.Unmarshal` call.

I can make the implementation of `UnmarshalJSON` inherently strict, but (I think) it requires unpacking into `json.RawMessage` and examining keys explicitly.  I'm not familiar with that at present.  In either event, I think it is a complex enough of an issue that it could be its own PR, if a temporary regression is acceptable.

Reply:
Thanks for the detailed investigation.

Is `UnmarshalJSON` the only way to do custom unmarshalling? Could you instead add `yaml:...` field tags to the matcher structs and perform the validation after the yaml has been loaded?

I think this wouldn't be too bad of a regression, as long as it is recovered in a timely manner. (Maybe consider creating an issue for visibility.)

Reply:
> Is UnmarshalJSON the only way to do custom unmarshalling? 

Kind of.  It is the ultimate unmarshaler for the `sigs.k8s.io/yaml` package.

> Could you instead add yaml:... field tags to the matcher structs and perform the validation after the yaml has been loaded?

Since the `sigs.k8s.io/yaml` package ultimately delegates to `json`, then the information would need to live in (non-existent) `json` tags.  Alternatively, we could switch back to the `gopkg.in/yaml` package.

The `gopkg.in` version has its pros and cons.  I went with the `sigs` version because it prevented us needing to tag every field with the corresponding value for config.  The `gopkg.in` version is very brittle to changes in capitalization, as we have in our `PackageRE`.  It expects keys to be strictly-lowercase, where the `sigs` version ends up with more case-agnostic unmarshaling.  But the `gopkg.in` version relies on a `UnmarshalYAML` interface, which provides a little bit more ease of use and would integrate more naturally with strict loading.

Migrading to the `gopkg.in` version would be another path forward.  But that is still a significant enough effort that would warrant its own PR.

Reply:
SG. I don't have a strong preference. I think it's fine to do this in a second PR.